internal/testtypes/builders.go

```diff
@@ -0,0 +1,530 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package testtypes
+
+import (
+	"fmt"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha4"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// ClusterBuilder holds the variables and objects required to build a clusterv1.Cluster.
+type ClusterBuilder struct {
+	namespace             string
+	name                  string
+	class                 clusterv1.ClusterClass
+	infrastructureCluster *unstructured.Unstructured
+	controlPlane          *unstructured.Unstructured
+}
+
+// NewClusterBuilder returns a ClusterBuilder with the given name and namespace.
+func NewClusterBuilder(namespace, name string) *ClusterBuilder {
+	return &ClusterBuilder{
+		namespace: namespace,
+		name:      name,
+	}
+}
+
+// WithInfrastructureCluster adds the passed InfrastructureCluster to the ClusterBuilder.
+func (c *ClusterBuilder) WithInfrastructureCluster(t *unstructured.Unstructured) *ClusterBuilder {
+	c.infrastructureCluster = t
+	return c
+}
+
+// WithControlPlane adds the passed ControlPlane to the ClusterBuilder.
+func (c *ClusterBuilder) WithControlPlane(t *unstructured.Unstructured) *ClusterBuilder {
+	c.controlPlane = t
+	return c
+}
+
+// WithClass adds the passed ClusterClass Name to the ClusterBuilder.
+func (c *ClusterBuilder) WithClass(class clusterv1.ClusterClass) *ClusterBuilder {
```

Comment:
```suggestion
func (c *ClusterBuilder) WithClusterClass(class clusterv1.ClusterClass) *ClusterBuilder {
```
given that we are always using full names.

Also in the future might be this must be transformed in a nested Topology builder